src/magic_folder/_endpoint_parser.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2020 Least Authority TFA GmbH
+# See COPYING for details.
+
+"""
+Turn Twisted endpoint description strings into HTTP(S) URLs.
+"""
+
+from hyperlink import (
+    URL,
+)
+
+def endpoint_description_to_http_api_root(endpoint_description):
+    """
+    Parse a Twisted endpoint description string and return a ``DecodedURL`` an
+    HTTP client could use to talk to an HTTP server listening on that
+    endpoint.
+
+    This currently supports only **tcp** and **ssl** endpoints.
+
+    :param unicode endpoint_description: The endpoint description string.
+
+    :return DecodedURL: A URL for reaching the given endpoint.
+    """
+    parts = endpoint_description.split(u":")
+    return _ENDPOINT_CONVERTERS[parts[0].lower()](parts[1:])
+
+
+def _tcp_to_http_api_root(parts):
+    """
+    Construct an HTTP URL.
+    """
+    port, host = _get_tcpish_parts(parts)
+    return URL(
+        scheme=u"http",
+        host=host,
+        port=port,
+    ).get_decoded_url()
+
+
+def _ssl_to_https_api_root(parts):
+    """
+    Construct an HTTPS URL.
+    """
+    port, host = _get_tcpish_parts(parts)
+    return URL(
+        scheme=u"https",
+        host=host,
+        port=port,
+    ).get_decoded_url()
+
+
+def _get_tcpish_parts(parts):
+    """
+    Split up the details of an endpoint with a host and port number (like
+    **tcp** and **ssl**).
+    """
+    port_number = int(parts.pop(0))
+    kwargs = dict(part.split(u"=", 1) for part in parts)
+    interface = kwargs.get(u"interface", u"127.0.0.1")
+    if interface == u"0.0.0.0":
+        interface = u"127.0.0.1"
+    return port_number, interface
+
+
+_ENDPOINT_CONVERTERS = {
+    u"tcp": _tcp_to_http_api_root,
+    u"ssl": _ssl_to_https_api_root,
+}
```

Comment:
Perhaps error handling could be improved here?   As in, what happens when `endpoint_description` is not really an endpoint?

I just learned that `magic-folder migrate --listen-endpoint 6001 --node-directory ../path/to/node` accepts that non-endpoint ar argument without complaining.  Thus magic folder configuration will have just `6001` for `api_endpoint`, and passing that on will result in an error here.

Of course `migrate` too probably could some error checking...

Reply:
Yea, it would be good to reject that early.  `create_global_configuration` should probably reject it and then the migrate command could handle that exception and report something meaningful to the cli user.

By the time it makes it into the database, it should really be valid and no one else in the system should have to worry about validation.

Reply:
Filed https://github.com/LeastAuthority/magic-folder/issues/226